### PR TITLE
fix(dsv4): align PD compress state layout

### PR DESCRIPTION
**Pull request by ziang663**
## Motivation

DeepSeek V4 PD disaggregation can run with asymmetric speculative decoding settings: the prefill server may run without speculative decoding, while the decode server enables EAGLE/MTP.

42645f15-f864-48ec-bf63-4230d7487d1d

For DSV4, the compress-state ring layout depends on whether the speculative-compatible layout is used. Before this change, a prefill server selected the non-speculative compress-state layout whenever speculative decoding was not enabled locally. In PD disaggregation, this can be wrong when the decode server uses EAGLE/MTP, because the prefill and decode servers then use different compress-state layouts and state transfer can break.

This PR lets a PD prefill server follow the decode server's speculative setting for the DSV4 compress-state layout without enabling speculative decoding on the prefill server itself.

## Modifications

- Add `--dsv4-compress-state-layout` to explicitly control the DSV4 compress-state ring layout:
  - `auto`
  - `speculative`
  - `non-speculative`
- Add `--dsv4-pd-decode-speculative-algorithm` so a PD prefill server can know whether the decode server enables speculative decoding.
- Add `use_speculative_compress_state_layout()` and use it when the DSV4 memory pool computes the compress-state ring size.
- Add unit tests covering:
  - normal non-PD non-speculative layout
  - local speculative decoding layout
  - PD prefill/decode speculative-compatible fallback
  - prefill following decode-side speculative algorithm
  - explicit layout override
  - c4/c128 ring-size differences between speculative and non-speculative layouts

Target scenario:

```text
P server:
  --disaggregation-mode prefill
  no --speculative-algorithm
  --dsv4-pd-decode-speculative-algorithm EAGLE

D server:
  --disaggregation-mode decode
  --speculative-algorithm EAGLE
```

In this setup, both sides use the speculative-compatible DSV4 compress-state layout.

## Accuracy Tests

This PR only changes DSV4 compress-state layout selection and CLI plumbing. It does not change model forward math, kernels, sampling, or logits.

Unit tests passed:

```bash
PYTHONPATH=python python test/registered/unit/mem_cache/test_deepseek_v4_memory_pool.py
```

Result:

```text
Ran 4 tests in 0.000s

OK
```

ServerArgs regression tests passed:

```bash
PYTHONPATH=python python test/registered/unit/server_args/test_server_args.py
```

Result:

```text
Ran 48 tests in 1.075s

OK
```

Manual layout check for the target PD setting:

```text
P_layout_spec True
D_layout_spec True
P_c4_ring 16 D_c4_ring 16
P_c128_ring 256 D_c128_ring 256
```

## Speed Tests and Profiling

No speed benchmark was run. This change only affects DSV4 compress-state layout selection during memory pool setup and does not add work to the model forward path.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### CODE REVIEW

**Comment by ziang663** [comment]
hi, could you please add the run-ci label or trigger CI for this PR? Thanks!
@merrymercy @yizhang2077

**Comment by xiezhq-hermann** [comment]
@hnyls2002 @ispobock @ShangmingCai can you take a look at this one?
